url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/process/DefaultProcessModel.java

```diff
@@ -0,0 +1,485 @@
+		return (objects.contains(obj));
+	}
+	
+	@Override
+	public String getName() {
+		return name;
+	}
+
+	@Override
+	public StartNode getStartNode() {
+		return startNode;
+	}
+
+	@Override
+	public Set<Trace> getDistinctTraces() {
+		Set<Trace> traces = getTracesFromNode(startNode, maxTraceLength);
+		return traces;
+	}
+
+	// TODO What about loops?
+	private Set<Trace> getTracesFromNode(ProcessNode node, int steps) {
+		Set<Trace> traces = new HashSet<Trace>();
+		
+		if (node instanceof EndNode || steps == 0) {
+			Trace newTrace = new Trace();
+			traces.add(newTrace);
+			return traces;
+		}
+		
+		if (node instanceof ActivityNode) {
+			return getTracesFromActivityNode((ActivityNode) node, steps);
+		}
+		if (node instanceof AndSplit) {
+			AndJoin andJoin = this.getMatchingAndJoin((AndSplit) node, steps-1);
+			Set<Trace> partTraces = this.getTracesInAndBlock((AndSplit) node, andJoin, steps);
+			Set<Trace> nextTraces = getTracesFromNode(andJoin, steps-1);
+			for (Trace t1 : partTraces) {
+				for (Trace t2 : nextTraces) {
+					Trace t = new Trace(t1.getActivities());
+					t.addAllActivities(t2.getActivities());
+					traces.add(t);
+				}
+			}
+		}
+		for (ProcessEdge edge : this.outgoingEdgesOf(node)) {
+			Set<Trace> outTraces = getTracesFromNode(this.getEdgeTarget(edge), steps-1);
+			traces.addAll(outTraces);
+		}
+		return traces;
+	}
+	
+	private Set<Trace> getTracesFromActivityNode(ActivityNode actNode, int steps) {
+		Set<Trace> traces = new HashSet<Trace>();
+		
+		Activity myAct = actNode.getActivity();
+		
+		if (this.outDegreeOf(actNode) == 0) {
+			Trace newTrace = new Trace();
+			newTrace.addActivity(myAct);
+			traces.add(newTrace);
+			return traces;
+		}
+		
+		for (ProcessEdge edge : this.outgoingEdgesOf(actNode)) {
+			Set<Trace> outTraces = getTracesFromNode(this.getEdgeTarget(edge), steps-1);
+			for (Trace outTrace : outTraces) {
+				Trace newTrace = new Trace();
+				newTrace.addActivity(myAct);
+				newTrace.addAllActivities(outTrace.getActivities());
+				traces.add(newTrace);
+			} 
+		}
+		return traces;
+	}
+	
+	
+	private Set<Trace> getTracesInAndBlock(AndSplit split, AndJoin andJoin, int steps) {
+		Set<Trace> traces = new HashSet<Trace>();
+		
+//		if (this.outDegreeOf(split) != 2) {
+//			throw new UnsupportedOperationException("And-blocks with more then two branches " +
+//				"are currently not supported.");
+//		}
+		
+		Map<Integer, Trace> branches = new HashMap<Integer, Trace>();
+		int index = 0;
+		for (ProcessEdge edge : this.outgoingEdgesOf(split)) {
+			index++;
+			Set<Trace> outTraces = getTracesFromNodeToNode(this.getEdgeTarget(edge), andJoin, steps-1);
+			if (outTraces.size() > 1) {
+				throw new UnsupportedOperationException("XorSplits in And-blocks are currently not supported");
+			}
+			branches.put(index, outTraces.iterator().next());
+		}
+		
+		List<Activity> activities = new ArrayList<Activity>();
+		for (Integer key : branches.keySet()) {
+			Trace t = branches.get(key);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.